### PR TITLE
Refactor logging to focus on simple STDOUT/STDERR

### PR DESCRIPTION
**Pull request by nathanleclaire**
This also lays the foundation for the possibility of log drivers in the
future, if it is decided that is a direction to pursue.

Closes https://github.com/docker/machine/issues/967.

`log.WithFields` still works pretty similarly ;D

cc @ehazlett @hairyhenderson @sthulb @bfirsh PTAL

Signed-off-by: Nathan LeClaire <nathan.leclaire@gmail.com>

### CODE REVIEW

**Comment by nathanleclaire** [comment]
### Concerns:

1. `github.com/Sirupsen/logrus` is still in `Godeps` - I don't really mind keeping it there in case we end up wanting to write a "driver" for it or use again for something in the future, but let me know your thoughts.  If I understand correctly removing it won't save us space on a repo clone or anything due to git tracking all of the history.
2.  I haven't implemented _all_ the methods that `logrus` originally supported - just the ones we have used - I definitely think there is an argument for moving over to only using only `log.*`, `log.*f` and killing `log.*ln` entirely.  See the interface if you are confused what I am talking about.

File: log/log.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	"os"
+	"strconv"
+)
+
+// Why the interface?  We may only want to print to STDOUT and STDERR for now,
+// but it won't neccessarily be that way forever.  This interface is intended
+// to provide a "framework" for a variety of different logging types in the
+// future (log to file, log to logstash, etc.) There could be a driver model
+// similar to what is done with OS or machine providers.
+type Logger interface {
```
Comment (nathanleclaire) [inline log/log.go]: this is the interface I am referencing in that "concerns" comment

**Comment by nathanleclaire** [comment]
investigating travis build

**Comment by sthulb** [comment]
I'd like to see logrus be removed. Cleaner working tree and less for travis to do (it uses shallow clones)

**Comment by nathanleclaire** [comment]
> I'd like to see logrus be removed. Cleaner working tree and less for travis to do (it uses shallow clones)

Ah yeah, that's a good reason.

**Comment by hairyhenderson** [comment]
Just a couple nitpicks:
- I would've expected to see `log.go` removed - is that missing from the commit (or am I blind?)
- I also would've expected to see `main.go`'s reference to `initLogging` (~ln19) removed

Otherwise, looks good :)

**Comment by ehazlett** [comment]
Overall LGTM.  +1 for removing logrus -- we can always add it back if needed.

**Comment by nathanleclaire** [comment]
> I would've expected to see log.go removed - is that missing from the commit (or am I blind?)
> I also would've expected to see main.go's reference to initLogging (~ln19) removed

Nope, it's not there - wanted to see the general sentiment around removing logrus completely was and it looks like everyone wants to, so I will put in those changes soon.

**Comment by nathanleclaire** [comment]
Update: I've removed the files which reference logrus in the main package, but unfortunately `github.com/docker/docker/utils` which gets imported by a few of our modules expects logrus to be present, so I haven't removed that from Godep for now.

**Comment by sthulb** [comment]
@nathanleclaire We can port docker/utils to machine, we only use one function out of the package.

File: main.go [hunk 2]
```diff
@@ -4,7 +4,6 @@ import (
 	"os"
 	"path"
 
-	log "github.com/Sirupsen/logrus"
```
Comment (hairyhenderson) [inline main.go]: You'll need to also import `"github.com/docker/machine/log"` here (`log` is referenced on ln 75)

**Comment by hairyhenderson** [comment]
@nathanleclaire & @sthulb - it looks like `docker/utils` is used in 3 places, and in 2 of the places it's referenced by an unreferenced unexpected function (`generateVMName` in azure.go and vcloudair.go). 

The last place is openstack.go, where it references `utils.GenerateRandomID`. I'm pretty sure we could port that usage, though honestly I'm not even sure why the KeyPair _needs_ to have a random suffix... But that's another issue.

**Comment by nathanleclaire** [comment]
OK I've removed `logrus`, moved the two functions that we depend on to `machine/utils/utils.go`, removed `docker/docker/api` and `/docker/docker/utils` deps (we were only using api in once place, and it also depends on `logrus`).  Seems to build and work fine, such a large change makes me a little nervous so maybe in the future we should do it in separate (follow-up) PRs.

**Comment by nathanleclaire** [comment]
Update:

1. Rebased
2. Moved `TerminalLogger` into its own package which `github.com/docker/machine/log` imports.  This will allow us to import different log drivers selectively if we want. For instance, in the drivers, we could change the import to something like `log "github.com/docker/machine/log/driver"`, which would allow the logger for the drivers to act slightly different than the default (prepend the machine name to the output, log to disk if there are errors, etc.).  It's also not infeasible to imagine the default logger as configurable via flag as well, so that for instance there could be an option to log with color in the output.
3. Filed an issue to remove the remaining `log.*ln` usages.

I think I'll make an issue to update the docs so that the output looks correct as well.

@ehazlett PTAL

**Comment by ehazlett** [comment]
reviewed, tested  - works as expected.  oh and yay for the windows side!!!

2015-04-30-154025_882x283_scrot

**Comment by ehazlett** [comment]
LGTM

**Comment by nathanleclaire** [comment]
:+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1:

**Pull request merged**